src/main/java/org/chameleoncloud/api/admin/UserGroupRolesResource.java

```diff
@@ -0,0 +1,320 @@
+    }
+
+    private void grantAdminPermissions(GroupModel group, GroupModel adminSubGroup) {
+	grantPermission("manage", group, adminSubGroup, UserGroupRoles.ADMIN);
+	grantPermission("manage-members", group, adminSubGroup, UserGroupRoles.ADMIN);
+	grantPermission("manage-membership", group, adminSubGroup, UserGroupRoles.ADMIN);
+    }
+
+    @GET
+    @Path("{user}")
+    @NoCache
+    @Produces(MediaType.APPLICATION_JSON)
+    public Response getUserGroupRoles(@QueryParam("group") String groupId, @PathParam("user") String userId) {
+
+	UserModel user = session.users().getUserById(userId, realm);
+	if (user == null) {
+	    return Response.status(Response.Status.NOT_FOUND).entity("user not found!").build();
+	}
+
+	Set<GroupModel> userGroups = user.getGroups();
+
+	if (groupId != null) {
+	    GroupModel group = session.realms().getGroupById(groupId, realm);
+	    Set<String> filterGroups = new HashSet<String>();
+	    filterGroups.add(groupId);
+	    for (GroupModel g : group.getSubGroups()) {
+		filterGroups.add(g.getId());
+	    }
+	    userGroups = userGroups.stream().filter(g -> filterGroups.contains(g.getId())).collect(Collectors.toSet());
+	}
+
+	Set<String> userGroupIds = new HashSet<String>();
+	// lazily grant permissions to user groups
+	for (GroupModel userGroup : userGroups) {
+	    String userGroupName = userGroup.getName();
+	    if (isRootGroup(userGroup)) {
+		grantMemberPermissions(userGroup);
+	    } else {
+		if (userGroupName.endsWith(UserGroupRoles.ADMIN_GROUP_SUFFIX)) {
+		    grantAdminPermissions(userGroup.getParent(), userGroup);
+		} else if (userGroupName.endsWith(UserGroupRoles.MANAGER_GROUP_SUFFIX)) {
+		    grantManagerPermissions(userGroup.getParent(), userGroup);
+		}
+	    }
+	    userGroupIds.add(userGroup.getId());
+	}
+
+	Set<UserGroupRoles> result = new HashSet<UserGroupRoles>();
+	for (GroupModel group : userGroups) {
+
+	    if (isRootGroup(group)) {
+
+		UserGroupRoles ugr = new UserGroupRoles();
+		ugr.setGroupId(group.getId());
+		ugr.setGroupName(group.getName());
+		SortedMap<Integer, String> policies = new TreeMap<Integer, String>();
+		Set<String> scopes = new HashSet<>();
+
+		for (String permissionName : groupPermission.getPermissions(group).keySet()) {
+		    Policy permission = getPolicyByPermissionName(permissionName, group);
+		    SortedMap<Integer, String> groupPolicies = getPolicies(permission, userGroupIds);
+		    if (!groupPolicies.isEmpty()) {
+			policies.putAll(groupPolicies);
+			scopes.addAll(getPermissionScopes(permission));
+		    }
+		}
+
+		ugr.setPolicy(policies.get(policies.firstKey()));
+		ugr.setScopes(scopes);
+		result.add(ugr);
+	    }
+	}
+
+	return Response.status(Response.Status.OK).entity(result).build();
+    }
+
+    @PUT
+    @Path("{user}")
+    @Consumes(MediaType.APPLICATION_JSON)
+    @Produces(MediaType.APPLICATION_JSON)
+    public Response setUserGroupRole(UserGroupRoles userGroupRoles, @PathParam("user") String userId) {
```

Comment:
how does `userGroupRoles` get filled in here? Is this some automagic thing that happens to the request JSON body?

Reply:
yes. fill in the json body with the class variables. the ones that are not provided will be null.